zcash_client_sqlite/src/wallet.rs

```diff
@@ -1701,25 +1786,20 @@ pub(crate) fn put_received_transparent_utxo<P: consensus::Parameters>(
     let address_str = output.recipient_address().encode(params);
     let account_id = conn
         .query_row(
-            "SELECT account FROM addresses WHERE cached_transparent_receiver_address = :address",
+            "SELECT account_id FROM addresses WHERE cached_transparent_receiver_address = :address",
             named_params![":address": &address_str],
-            |row| row.get::<_, u32>(0),
+            |row| Ok(AccountId(row.get(0)?)),
         )
         .optional()?;
 
     let utxoid = if let Some(account) = account_id {
-        put_legacy_transparent_utxo(
-            conn,
-            params,
-            output,
-            AccountId::try_from(account).map_err(|_| SqliteClientError::AccountIdOutOfRange)?,
-        )?
+        put_legacy_transparent_utxo(conn, params, output, account)?
     } else {
         // If the UTXO is received at the legacy transparent address, there may be no entry in the
         // addresses table that can be used to tie the address to a particular account. In this
         // case, we should look up the legacy address for account 0 and check whether it matches
         // the address for the received UTXO, and if so then insert/update it directly.
-        let account = AccountId::ZERO;
+        let account = AccountId(0);
```

Comment:
Bug: This is intentionally meant to be referring to ZIP 32 account index 0. If a new wallet is created where the first account is either imported or not ZIP 32 account 0, this will break.

You need to search the accounts table for every account that has ZIP 32 account index 0, and check each of their legacy transparent addresses against the address for the received UTXO. If none match, then the existing `SqliteClientError::AddressNotRecognized` can be returned; if more than one match, then you need to return `SqliteClientError::CorruptedData`.

Reply:
Very good catch.

Reply:
By "_legacy_ transparent addresses", does _legacy_ imply something specific (e.g. a particular address index)?

Reply:
I'm also wondering why this more exhaustive search would be limited to just zip32 account index 0 accounts. Shouldn't this test the default transparent address for _every_ account in the accounts table?

Reply:
Yes: it implies diversifier index zero.

The problem here is caused by a confluence of issues:
- 50% of diversifier indices are not valid for Sapling, so the "default" UA containing a Sapling receiver corresponds to the smallest diversifier index that can generate a Sapling address.
- To ease migration of wallets from Sapling-only or transparent-plus-Sapling (specifically to avoid additional scanning load if a separate Sapling account was created), the UA derivation pathways use the existing ZIP 32 and BIP 44 pathways.
- To ensure that UAs can be diversified, the diversifier index is mapped to the transparent "address index".

This means that for wallets that either were transparent-only or transparent-plus-Sapling prior to UAs, the default BIP 44 transparent address at address index 0 that was likely being used before, may or may not be in the default UA (it will be for some accounts, and won't be for others). In that case, the `cached_transparent_receiver_address` column in the `addresses` table (which contains the transparent receiver from the UA) will only contain the address index 0 transparent address in 50% of wallets.

This code is handling the edge case, by looking up which account's address index 0 transparent address matches the UTXO (if any), and then inserting the UTXO into the wallet DB using that account.

Reply:
Oh, does the "default diversifier index might not be 0 but the default t-addr is at index 0" issue only affect ZIP 32 accounts with *account index* 0? Why? Is it just because the wallets that did this were single-account wallets?

Reply:
> Is it just because the wallets that did this were single-account wallets?

Yes; that's the problem that was fixed in #685. However, our API (at least in current `main`) will return the t-addr at index 0 from all calls to `WalletRead::get_transparent_receivers(account)`, so #685 was technically just a partial fix to the general problem (which we also generally never encountered because anyone using our SDK with multiple accounts was not doing so to import a prior transparent wallet that used multiple accounts, otherwise we'd have received bug reports about this).

The way I'm going to fix this will implicitly mean we check all accounts correctly.

Reply:
Here's the fix (I can't suggest it because it covers lines outside the diff):
```rust
    if let Some(account) = account_id {
        Ok(put_legacy_transparent_utxo(conn, params, output, account)?)
    } else {
        // If the UTXO is received at the legacy transparent address (at BIP 44 address
        // index 0 within its particular account, which we specifically ensure is returned
        // from `get_transparent_receivers`), there may be no entry in the addresses table
        // that can be used to tie the address to a particular account. In this case, we
        // look up the legacy address for each account in the wallet, and check whether it
        // matches the address for the received UTXO; if so, insert/update it directly.
        get_account_ids(conn)?
            .into_iter()
            .find_map(
                |account| match get_legacy_transparent_address(params, conn, account) {
                    Ok(Some((legacy_taddr, _))) if &legacy_taddr == output.recipient_address() => {
                        Some(
                            put_legacy_transparent_utxo(conn, params, output, account)
                                .map_err(SqliteClientError::from),
                        )
                    }
                    Ok(_) => None,
                    Err(e) => Some(Err(e)),
                },
            )
            // The UTXO was not for any of the legacy transparent addresses.
            .unwrap_or(Err(SqliteClientError::AddressNotRecognized(
                *output.recipient_address(),
            )))
    }
```